frontend/src/app/modules/matches/matches-computations.js

```diff
@@ -68,3 +77,104 @@ export const computeWinRatio = (playerId, playerMatches) => {
   return (playerMatches.length > 0) ? (wonMatchesCount / playerMatches.length) : 0
 }
+
+export const computeDays = matchChanges => {
+  let worstDayKey = ''
+  let bestDayKey = ''
+  const daysMap = {}
+  for (const match of matchChanges) {
+    const key = match.date.toLocaleDateString()
+    const value = Number(match.ratingChangeString)
+    if (!Object.hasOwnProperty.call(daysMap, key)) {
+      daysMap[key] = {
+        date: key,
+        value,
+      }
+    } else {
+      daysMap[key]['value'] += value
+    }
+
+    bestDayKey = updateMaxFromMap(daysMap[bestDayKey] ?? {}, 'value', daysMap[key]).date
+    worstDayKey = updateMinFromMap(daysMap[worstDayKey] ?? {}, 'value', daysMap[key]).date
+  }
+
+  return {
+    bestDay: daysMap[bestDayKey],
+    worstDay: daysMap[worstDayKey],
+  }
+}
+
+export const computeTeammates = (playerId, playerMatches) => {
+  const teammates = findSignificanPlayers(playerId, playerMatches, getTeamMates)
+  return {
+    mostFrequentTeammate: teammates.mostFrequentPlayer,
+    leastFrequentTeammate: teammates.leastFrequentPlayer,
+    mostSuccessTeammate: teammates.mostSuccessPlayer,
+    leastSuccessTeammate: teammates.leastSuccessPlayer,
+  }
+}
+
+export const computeOpponents = (playerId, playerMatches) => {
+  const opponents = findSignificanPlayers(playerId, playerMatches, getOpponents)
+  return {
+    mostFrequentOpponent: opponents.mostFrequentPlayer,
+    leastFrequentOpponent: opponents.leastFrequentPlayer,
+    mostSuccessOpponent: opponents.mostSuccessPlayer,
+    leastSuccessOpponent: opponents.leastSuccessPlayer,
+  }
+}
+
+const findSignificanPlayers = (playerId, playerMatches, playersProvider) => {
+  let mostFrequentPlayer = {}
+  let leastFrequentPlayer = {}
+  let mostSuccessPlayer = {}
+  let leastSuccessPlayer = {}
+  const playersMap = {}
+  for (const match of playerMatches) {
+    const players = playersProvider(playerId, match)
+    for (const player of players) {
+      if (!Object.hasOwnProperty.call(playersMap, player.id)) {
+        playersMap[player.id] = {
+          player: player,
+          matches: 1,
+          wins: Number(didPlayerWin(playerId, match)),
+          winRatio: Number(didPlayerWin(playerId, match)),
+        }
+      } else {
+        ++playersMap[player.id]['matches']
+        playersMap[player.id]['wins'] += Number(didPlayerWin(playerId, match))
+        playersMap[player.id]['losses'] = playersMap[player.id]['matches'] - playersMap[player.id]['wins']
+      }
+
+      mostFrequentPlayer = updateMaxFromMap(mostFrequentPlayer, 'matches', playersMap[player.id])
+      leastFrequentPlayer = updateMinFromMap(leastFrequentPlayer, 'matches', playersMap[player.id])
+      mostSuccessPlayer = updateMaxFromMap(mostSuccessPlayer, 'wins', playersMap[player.id])
+      leastSuccessPlayer = updateMaxFromMap(leastSuccessPlayer, 'losses', playersMap[player.id])
+    }
+  }
+
+  return {
```

Comment:
what if there are not wins or no loses or no matches? is it ok to return empty objects? I am afraid links can break

Reply:
Fixed by providing default values [here](https://github.com/salsita/foosball-rating/pull/114/commits/3d398530cdf75f980e6344e885e4f18dd919311a#diff-304dca28d92d292252e86b7e83fc1569)